storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/handler/LogviewerLogPageHandler.java

```diff
@@ -0,0 +1,467 @@
+import org.apache.storm.daemon.logviewer.utils.ResourceAuthorizer;
+import org.apache.storm.daemon.logviewer.utils.WorkerLogs;
+import org.apache.storm.daemon.utils.StreamUtil;
+import org.apache.storm.daemon.utils.UrlBuilder;
+import org.apache.storm.ui.InvalidRequestException;
+import org.apache.storm.ui.UIHelpers;
+import org.apache.storm.utils.ConfigUtils;
+import org.apache.storm.utils.ServerUtils;
+import org.apache.storm.utils.Utils;
+import org.jooq.lambda.Unchecked;
+
+public class LogviewerLogPageHandler {
+    private final String logRoot;
+    private final String daemonLogRoot;
+    private final WorkerLogs workerLogs;
+    private final ResourceAuthorizer resourceAuthorizer;
+
+    /**
+     * Constructor.
+     *
+     * @param logRoot root worker log directory
+     * @param daemonLogRoot root daemon log directory
+     * @param workerLogs {@link WorkerLogs}
+     * @param resourceAuthorizer {@link ResourceAuthorizer}
+     */
+    public LogviewerLogPageHandler(String logRoot, String daemonLogRoot,
+                                   WorkerLogs workerLogs,
+                                   ResourceAuthorizer resourceAuthorizer) {
+        this.logRoot = logRoot;
+        this.daemonLogRoot = daemonLogRoot;
+        this.workerLogs = workerLogs;
+        this.resourceAuthorizer = resourceAuthorizer;
+    }
+
+    /**
+     * Enumerate worker log files for given criteria.
+     *
+     * @param user username
+     * @param port worker's port, null for all workers
+     * @param topologyId topology ID, null for all topologies
+     * @param callback callback for JSONP
+     * @param origin origin
+     * @return list of worker logs for given criteria
+     */
+    public Response listLogFiles(String user, Integer port, String topologyId, String callback, String origin) throws IOException {
+        List<File> fileResults = null;
+        if (topologyId == null) {
+            if (port == null) {
+                fileResults = workerLogs.getAllLogsForRootDir();
+            } else {
+                fileResults = new ArrayList<>();
+
+                File[] logRootFiles = new File(logRoot).listFiles();
+                if (logRootFiles != null) {
+                    for (File topoDir : logRootFiles) {
+                        File[] topoDirFiles = topoDir.listFiles();
+                        if (topoDirFiles != null) {
+                            for (File portDir : topoDirFiles) {
+                                if (portDir.getName().equals(port.toString())) {
+                                    fileResults.addAll(DirectoryCleaner.getFilesForDir(portDir));
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        } else {
+            if (port == null) {
+                fileResults = new ArrayList<>();
+
+                File topoDir = new File(logRoot + Utils.FILE_PATH_SEPARATOR + topologyId);
```

Comment:
nit:  I know this is an exact translation, but we could just use `new File(logRoot, topologyId)`